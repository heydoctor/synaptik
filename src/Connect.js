import React from 'react';
import PropTypes from 'prop-types';
import Context from './Context';

class ConnectConsumer extends React.PureComponent {
  unsubscribe = null;
  state = this.getObservedState();

  static propTypes = {
    select: PropTypes.func.isRequired,
    lifecycle: PropTypes.shape({
      didMount: PropTypes.func,
      didUpdate: PropTypes.func,
      willUnmount: PropTypes.func,
    }),
  };

  static defaultProps = {
    lifecycle: {},
  };

  componentDidMount() {
    let { lifecycle, synapse } = this.props;

    this.unsubscribe = synapse.subscribe(() => {
      this.setState(this.getObservedState());
    });

    if (lifecycle.didMount) lifecycle.didMount(...this.getArgs());
  }

  componentDidUpdate() {
    let { lifecycle } = this.props;
    if (lifecycle.didUpdate) lifecycle.didUpdate(...this.getArgs());
  }

  componentWillUnmount() {
    this.unsubscribe();

    let { lifecycle } = this.props;
    if (lifecycle.willUnmount) lifecycle.willUnmount(...this.getArgs());
  }

  getArgs() {
    let { synapse, lifecycle, select, ...props } = this.props;
    return [synapse.stores, props];
  }

  getObservedState() {
    let state = this.props.select(...this.getArgs());
    if (typeof state !== 'object') {
      throw new Error('Expected `select` to return an object.');
    }
    return state;
  }

  render() {
    let { children, synapse, ...props } = this.props;

    return children(this.state, props);
  }
}

export default function Connect(props) {
  return (
    <Context.Consumer>
      {synapse => <ConnectConsumer {...props} synapse={synapse} />}
    </Context.Consumer>
  );
}

export function connect(selector, lifecycle = {}) {
  return Component => {
    return props => (
      <Connect {...props} select={selector} lifecycle={lifecycle}>
        {state => <Component {...state} {...props} />}
      </Connect>
    );
  };
}
