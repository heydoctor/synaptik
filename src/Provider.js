import React from 'react';
import PropTypes from 'prop-types';
import Context from './Context';
import Synapse from './Synapse';

export default class Provider extends React.Component {
  static propTypes = {
    synapse: PropTypes.shape(),
    stores: PropTypes.shape(),
  };

  constructor(props, context) {
    super(props, context);

    this.synapse = props.synapse
      ? props.synapse
      : new Synapse(props.stores, { logger: props.logger });
  }

  render() {
    return (
      <Context.Provider value={this.synapse}>
        {this.props.children}
      </Context.Provider>
    );
  }
}
