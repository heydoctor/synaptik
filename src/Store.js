export default class Store {
  constructor(id, synapse) {
    if (!id) throw new Error('Store requires an id');
    if (!synapse) throw new Error('Store requires a synapse instance');

    this.id = id;
    this.synapse = synapse;
    this.state = {};
  }

  setState(updater, { log = true } = {}) {
    let updates = typeof updater === 'function' ? updater(this.state) : updater;
    this.state = { ...this.state, ...updates };
    this.synapse.updateState(this.id, this.state, { log });
  }
}
