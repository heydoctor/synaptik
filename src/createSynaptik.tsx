import React, {
  useContext,
  useState,
  useEffect,
  useRef,
  useCallback,
  forwardRef,
  ComponentType,
} from 'react';
import shallowEqual from 'shallowequal';

type SubscriptionFn = (state: any) => any;

export type ConstructorMap = Record<string, new (...args: any[]) => any>;

export type StoreMap<T extends ConstructorMap> = {
  [P in keyof T]?: InstanceType<T[P]>;
};

export type Synapse<Stores extends ConstructorMap> = {
  subscribe: (subscription: SubscriptionFn) => () => void;
  notify: () => void;
  updateState: (id: string, newState: any) => void;
  getState: () => {
    [P in keyof Stores]?: InstanceType<Stores[P]>['state'];
  };
  stores: StoreMap<Stores>;
};

export interface SynaptikInstance<Stores extends ConstructorMap> extends Synapse<Stores> {
  Provider: ComponentType;
  useSynapse: <TSelected>(
    selector: (stores: StoreMap<Stores>) => TSelected,
    dependencies?: any[]
  ) => TSelected;
  connect: <TSelected>(
    selector: (stores: StoreMap<Stores>) => TSelected,
    dependencies?: any[]
  ) => (Component: ComponentType) => any;
}

export function createSynaptik<T extends ConstructorMap>(passedStores: T): SynaptikInstance<T> {
  const state: {
    [P in keyof T]?: InstanceType<T[P]>['state'];
  } = {};
  const stores: StoreMap<T> = {};
  const subscriptions: Map<number, SubscriptionFn> = new Map();
  let currentSubscriptionId = 0;

  function notify() {
    for (const fn of subscriptions.values()) {
      fn(state);
    }
  }

  function subscribe(subscription: SubscriptionFn) {
    if (typeof subscription !== 'function') {
      throw new Error('Subscription must be a function');
    }

    const id = ++currentSubscriptionId;
    subscriptions.set(id, subscription);

    return () => {
      subscriptions.delete(id);
    };
  }

  function getState() {
    return state;
  }

  function updateState(storeId: keyof T, newState: any) {
    state[storeId] = newState;
    notify();
  }

  const synapse = {
    stores,
    getState,
    updateState,
    subscribe,
    notify,
  };

  Object.entries(passedStores).forEach(([id, Store]) => {
    const store = new Store(id, synapse);
    stores[id as keyof T] = store;
    updateState(id, store.state);
  });

  const Context = React.createContext(synapse);

  const Provider: React.FC = ({ children }) => (
    <Context.Provider value={synapse}>{children}</Context.Provider>
  );

  function useSynapse<TSelected>(
    selector: (stores: StoreMap<T>) => TSelected,
    dependencies: any[] = []
  ) {
    const { subscribe } = useContext(Context);
    const select = useCallback(() => selector(stores), []);
    const [state, setState] = useState(select());

    // By default, our effect only fires on mount and unmount, meaning it won't see the
    // changes to state, so we use a mutable ref to track the current value
    const stateRef = useRef(state);

    useEffect(() => {
      // Helps to avoid running stale listeners after unmount
      let isUnsubscribed = false;

      const maybeUpdateState = () => {
        if (isUnsubscribed) {
          return;
        }

        const nextState = select();

        // Checking referential equality grants perf boost if selector is memoized
        if (nextState === stateRef.current || shallowEqual(nextState, stateRef.current)) {
          return;
        }

        stateRef.current = nextState;
        setState(nextState);
      };

      const unsubscribe = subscribe(maybeUpdateState);

      maybeUpdateState();

      return () => {
        unsubscribe();
        isUnsubscribed = true;
      };
    }, dependencies);

    return state;
  }

  function connect<TSelected>(
    selector: (stores: StoreMap<T>) => TSelected,
    dependencies: any[] = []
  ) {
    return (Component: ComponentType<any>) =>
      forwardRef((props, ref) => {
        const state = useSynapse(selector, dependencies);
        return <Component {...state} {...props} ref={ref} />;
      });
  }

  return {
    Provider,
    useSynapse,
    connect,
    ...synapse,
  };
}
