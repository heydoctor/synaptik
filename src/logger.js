const detailedDiff = require('deep-object-diff').detailedDiff;

const logger = (message, color, object) => {
  if (object && Object.keys(object).length) {
    console.log(
      `%c${message}`,
      `color: ${color}; font-weight: bold;\n`,
      object
    );
  }
};

export default function logStateChange(prevState, newState) {
  const diff = detailedDiff(prevState, newState);
  const groupName = `Vault Update`;

  console.groupCollapsed(groupName);
  logger('Added', '#18A76D', diff.added);
  logger('Updated', '#FDA429', diff.updated);
  logger('Deleted', '#f04545', diff.deleted);
  logger('Old state: ', '#555555', Object.assign({}, prevState));
  logger('New state: ', '#555555', Object.assign({}, newState))
  console.groupEnd(groupName);
};
