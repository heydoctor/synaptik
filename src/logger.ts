const detailedDiff = require('deep-object-diff').detailedDiff;

const logger = (message: string, color: string, object: any) => {
  if (object && Object.keys(object).length) {
    console.log(`%c${message}`, `color: ${color}; font-weight: bold;\n`, object);
  }
};

export default function logStateChange(prevState: any, newState: any) {
  const diff = detailedDiff(prevState, newState);

  console.groupCollapsed('Vault Update');
  logger('Added', '#18A76D', diff.added);
  logger('Updated', '#FDA429', diff.updated);
  logger('Deleted', '#f04545', diff.deleted);
  logger('Old state: ', '#555555', Object.assign({}, prevState));
  logger('New state: ', '#555555', Object.assign({}, newState));
  console.groupEnd();
}
