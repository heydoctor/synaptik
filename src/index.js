import Connect, { connect } from './Connect';
import Provider from './Provider';
import Store from './Store';
import useSynapse from './useSynapse';
import Synapse from './Synapse';
import logger from './logger';

export { Synapse, Provider, Connect, connect, Store, useSynapse, logger };
