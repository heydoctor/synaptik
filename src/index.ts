import { Store } from './Store';
import { Synapse } from './Synapse';
import { createSynaptik } from './createSynaptik';
import logger from './logger';

export { Store, Synapse, createSynaptik, logger };
