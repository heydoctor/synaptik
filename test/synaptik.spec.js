import * as synaptik from '../src';

class TestStore extends synaptik.Store {
  state = {
    counter: 0,
  };
}

let synapse;
beforeEach(() => {
  synapse = new synaptik.Synapse({});
});

describe('Synapse', () => {
  test('logger', () => {
    const logger = jest.fn();
    const synapse = new synaptik.Synapse({}, { logger });
    // logs by default
    synapse.updateState();
    expect(logger).toHaveBeenCalled();

    logger.mockReset();

    // ignore logs
    synapse.updateState(null, null, { log: false });
    expect(logger).not.toHaveBeenCalled();
  });

  test('updating state', () => {
    synapse.updateState('test', {
      counter: 1,
    });
    const newState = synapse.getState();
    expect(newState.test.counter).toEqual(1);
  });

  test('subscriptions', () => {
    const spy = jest.fn();

    // Subscribe to the synapse
    const subscription = synapse.subscribe(spy);
    synapse.updateState('test', {
      counter: 1,
    });
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(synapse.state);


    // Unsubscribe to the synapse
    subscription();

    synapse.updateState('test', {
      counter: 2,
    });

    expect(spy).toHaveBeenCalledTimes(1);
  });
});

describe('Store', () => {
  let store;

  beforeEach(() => {
    store = new TestStore('test', synapse);
  });

  test('requires id and synapse', () => {
    expect(() => new TestStore()).toThrowErrorMatchingSnapshot();
    expect(() => new TestStore('test')).toThrowErrorMatchingSnapshot();
  });

  test('setState returns a promise', () => {
    const promise = store.setState({});
    expect(typeof promise.then).toEqual('function');
  });

  test('setState can take an updater function', async () => {
    const spy = jest.fn().mockImplementationOnce(() => ({}));
    const promise = await store.setState(spy);
    expect(spy).toHaveBeenCalledWith(store.state);
  });

  test('setState calls synapse.updateState', async () => {
    const spy = jest.fn().mockImplementationOnce(() => ({}));
    synapse.updateState = jest.fn();
    await store.setState(spy);
    expect(synapse.updateState).toHaveBeenCalledWith(store.id, store.state, {
      log: true,
    });
  });
});
