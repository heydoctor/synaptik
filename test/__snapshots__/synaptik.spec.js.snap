// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store requires id and synapse 1`] = `"Store requires an id"`;

exports[`Store requires id and synapse 2`] = `"Store requires a synapse instance"`;
